src/oracle.py

```diff
@@ -0,0 +1,135 @@
+from time import sleep
+from typing import Optional, List
+
+from prometheus_client import start_http_server
+from web3 import Web3
+from web3_multi_provider import MultiProvider
+
+from src import variables
+from src.contracts import contracts
+from src.metrics.logging import logging
+from src.metrics.healthcheck_server import start_pulse_server, pulse
+from src.metrics.prometheus.basic import FINALIZED_EPOCH_NUMBER, SLOT_NUMBER
+
+from src.metrics.prometheus.basic import EXCEPTIONS_COUNT
+from src.modules.accounting import Accounting
+from src.modules.ejection import Ejector
+from src.modules.interface import OracleModule
+from src.protocol_upgrade_checker import wait_for_withdrawals
+from src.providers.beacon import BeaconChainClient, NoSlotsFound
+from src.providers.web3_middleware import add_requests_metric_middleware
+from src.web3_utils.typings import EpochNumber, SlotNumber
+from src.variables import DAEMON, WEB3_PROVIDER_URI, BEACON_NODE
+
+logger = logging.getLogger(__name__)
+
+
+class Oracle:
+    _last_finalized_epoch: EpochNumber = 0
+
+    def __init__(self, web3: Web3, beacon_chain_client: BeaconChainClient):
+        self._w3 = web3
+        self._beacon_chain_client = beacon_chain_client
+
+        self.modules: List[OracleModule] = [
+            Accounting(self._w3, self._beacon_chain_client),
+            Ejector(self._w3, self._beacon_chain_client),
+        ]
+
+    def _fetch_beacon_specs(self):
+        _, self.slots_per_epoch, self.seconds_per_slot, _ = contracts.oracle.functions.getBeaconSpec().call()
+        # _, self.slots_per_epoch, self.seconds_per_slot, _ = contracts.validator_exit_bus.functions.getBeaconSpec().call()
+
+    def run_as_daemon(self):
+        while True:
+            pulse()
+            logger.info({'msg': 'Run cycle.'})
+            self.run_cycle()
+
+    def run_cycle(self):
+        self._fetch_beacon_specs()
+        epoch = self._fetch_next_finalized_epoch()
+        logger.info({'msg': 'Get finalized epoch.', 'value': epoch})
+
+        try:
+            self.run_once(epoch)
+        except KeyboardInterrupt as error:
```

Comment:
I'm not python expert. Moreover, "KeyboardInterrupt" tells me about user's or system app termination.
I'd prefer to move this signal under "Unexpected exception." and rename it like "SIGTERM" (signal for termination)

Reply:
This default python exception